languages/swift/Flipper/Module.swift

```diff
@@ -0,0 +1,109 @@
+//
+//  Created by Harlan Haskins on 2/9/18.
+//
+
+import Foundation
+import CFlipper
+
+public struct StandardModuleFFI {
+  let moduleMetadata: _lf_module
+}
+
+public struct UserModuleFFI {
+  var moduleMetadata: _lf_module
+  let name: String
+
+  init(name: String, version: lf_version_t, crc: lf_crc_t, index: Int32) {
+    self.name = name
+    self.moduleMetadata = _lf_module(name: name,
+                                     description: nil,
+                                     version: version,
+                                     identifier: crc,
+                                     index: index,
+                                     device: nil,
+                                     data: nil,
+                                     psize: nil)
+  }
+}
+
+func buildLinkedList(_ args: [LFArg]) -> UnsafeMutablePointer<_lf_ll>? {
+  var argList: UnsafeMutablePointer<_lf_ll>? = nil
+  for arg in args {
+    lf_ll_append(&argList,
+                 UnsafeMutableRawPointer(bitPattern:
+                  UInt(truncatingIfNeeded: arg.asLFArg.value)), nil)
+  }
+  return argList
+}
+
+enum ModuleFFI {
+  case standard(StandardModuleFFI)
+  case user(UserModuleFFI)
+
+  var modulePtr: _lf_module {
+    switch self {
+    case .standard(let mod): return mod.moduleMetadata
+    case .user(let mod): return mod.moduleMetadata
+    }
+  }
+
+  func invoke(index: UInt8, args: [LFArg]) {
+    _ = invoke(index: index, args: args) as LFVoid
+  }
+
+  func push(index: UInt8, data: Data, args: [LFArg]) {
+    _ = push(index: index, data: data, args: args) as LFVoid
+  }
+
+  func pull(index: UInt8, data: inout Data, args: [LFArg]) {
+    _ = pull(index: index, data: &data, args: args) as LFVoid
+  }
+
+  func invoke<Ret: LFReturnable>(index: UInt8, args: [LFArg]) -> Ret {
+    var mod = modulePtr
+    let ret = lf_invoke(&mod, index, Ret.lfType.rawValue, buildLinkedList(args))
+    return Ret.init(lfReturn: ret)
+  }
+
+  func push<Ret: LFReturnable>(index: UInt8, data: Data, args: [LFArg]) -> Ret {
+    var mod = modulePtr
+    let ret = data.withUnsafeBytes {
+      lf_push(&mod, index, UnsafeMutablePointer(mutating: $0),
+              UInt32(data.count), buildLinkedList(args))
+    }
+    return Ret.init(lfReturn: ret)
+  }
+
+  func pull<Ret: LFReturnable>(index: UInt8, data: inout Data,
+                                args: [LFArg]) -> Ret {
+    var mod = modulePtr
+    let count = UInt32(data.count)
+    let ret = data.withUnsafeMutableBytes {
+      lf_pull(&mod, index, $0, count, buildLinkedList(args))
+    }
+    return Ret.init(lfReturn: ret)
+  }
+}
+
+public protocol UserModule {
+  static var name: String { get }
+  init(flipper: Flipper)
+  init(ffi: UserModuleFFI)
+  init()
+}
+
+public extension UserModule {
+  init(flipper: Flipper) {
+    var ffi = UserModuleFFI(name: Self.name, version: 0, crc: 0, index: 0)
+    lf_bind(&ffi.moduleMetadata, flipper.device)
+    self.init(ffi: ffi)
+  }
+}
+
+public protocol StandardModule {
```

Comment:
In Swift, are you able to pass around implementing types of protocols with their types erased? For example, in Java you can have this:

```java
interface Foo { }
class Bar implements Foo { }
class Baz implements Foo { }
List<Foo> foos = new ArrayList<>();
foos.add(new Bar());
foos.add(new Baz());
```

Whereas in Rust you cannot (perhaps a `&Trait` reference, but not by ownership). I'm considering removing the `StandardModule` trait from Rust because it may not really buy us anything. Since we only have a fixed number of modules we have to implement them for, we could decide to maintain API consistency just by convention rather than by constraining them to a trait (or protocol in this case) if that explicit constraint doesn't buy us anything.

If Swift does treat protocols like Java treats interfaces, then we have to decide whether there's really a use case for constraining them to a protocol. For example, is it a reasonable idea to load up a list with `StandardModule` objects and use that list to init them all at once? What I'm thinking of may look something like this in Java:

```java
List<StandardModule> modulesBeingUsed = Arrays.asList(
    new Gpio(),
    new Spi(),
    // etc.
);
for (StandardModule s : modulesBeingUsed) { s.init(); }
```

I think this universal representation of standard modules may stop being useful immediately beyond wanting to `init` several modules at once, since the rest of the module APIs are unique per module. i'd like to hear other thoughts on this though :)

---

I realize this may not be the best place for this question, since it doesn't really have to do specifically with the Swift bindings and I'm not criticizing them specifically, but this just made me think back to some unfinished thoughts I had about the module APIs in general.

Reply:
No no, this is a good place for this discussion. Yes, Swift does allow you to use protocols as 'existentials', so you could, say, have a `[StandardModule]`. In practice, I don't think there's much of a use case _at all_ for someone passing around modules, especially when they're type-erased. Since the protocol only declares initializers, there's no common API exposed on them once they've been created. I.e. a `[StandardModule]` gets you exactly 0 bits of functionality...